src/lib/openapi-to-json-schema.js

```diff
@@ -0,0 +1,414 @@
+ *
+ * And why use this over `@openapi-contrib/openapi-schema-to-json-schema`? Fortunately and unfortunately we've got a lot
+ * of API definitions in our database that aren't currently valid so we need to have a lot of bespoke handling for odd
+ * quirks, typos, and missing declarations that they've got.
+ *
+ * @link https://json-schema.org/draft/2019-09/json-schema-validation.html
+ * @link https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md
+ * @param {Object} data
+ * @param {Object} opts
+ * @param {Object[]} opts.prevSchemas - Array of parent schemas to utilize when attempting to path together examples.
+ * @param {String} opts.currentLocation - Current location within the schema -- this is a JSON pointer.
+ * @param {Object} opts.globalDefaults - Object containing a global set of defaults that we should apply to schemas that match it.
+ * @param {Boolean} opts.isPolymorphicAllOfChild - Is this schema the child of a polymorphic `allOf` schema?
+ */
+function toJSONSchema(data, opts = {}) {
+  const schema = { ...data };
+  const { prevSchemas, currentLocation, globalDefaults, isPolymorphicAllOfChild } = {
+    prevSchemas: [],
+    currentLocation: '',
+    globalDefaults: {},
+    isPolymorphicAllOfChild: false,
+    ...opts,
+  };
+
+  // If this schema contains a `$ref`, it's circular and we shouldn't try to resolve it. Just return and move along.
+  if (schema.$ref) {
+    return {
+      $ref: schema.$ref,
+    };
+  }
+
+  // If we don't have a set type, but are dealing with an `anyOf`, `oneOf`, or `allOf` representation let's run through
+  // them and make sure they're good.
+  ['allOf', 'anyOf', 'oneOf'].forEach(polyType => {
+    if (polyType in schema && Array.isArray(schema[polyType])) {
+      schema[polyType].forEach((item, idx) => {
+        const polyOptions = {
+          prevSchemas,
+          currentLocation: `${currentLocation}/${idx}`,
+          globalDefaults,
+          isPolymorphicAllOfChild: polyType === 'allOf',
+        };
+
+        // When `properties` or `items` are present alongside a polymorphic schema instead of letting whatever JSON
+        // Schema interpreter is handling these constructed schemas we can guide its hand a bit by manually transforming
+        // it into an inferred `allOf` of the `properties` + the polymorph schema.
+        if ('properties' in schema) {
+          schema[polyType][idx] = toJSONSchema({ allOf: [item, { properties: schema.properties }] }, polyOptions);
+        } else if ('items' in schema) {
+          schema[polyType][idx] = toJSONSchema({ allOf: [item, { items: schema.items }] }, polyOptions);
+        } else {
+          schema[polyType][idx] = toJSONSchema(item, polyOptions);
+        }
+      });
+    }
+  });
+
+  // If this schema is malformed for some reason, let's do our best to repair it.
+  if (!('type' in schema) && !isPolymorphicSchema(schema) && !isRequestBodySchema(schema)) {
```

Comment:
I really worry about modifying people's definitions as they're uploaded because when we offer it back to them it likely won't be the same as what they sent us.

Reply:
yeah of course, we would need the DL link to  be the original, and not the optimized one. the more I think of it though, the more I see a massive headache of rendering from a file different from what's uploaded. the transform seems safer.

Reply:
Agreed, and also saves us the headache of having to re-optimize definitions in the database when we fix bugs in this optimization/conversion process.